# nf-core/lncpipe: Output

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report and  LncpipeReporter, that summarizes the result at the end of the pipeline.


## Pipeline overview
The pipeline is built using [Nextflow](https://www.nextflow.io/)
Plz see a detailed illustration in our [literature](https://linkinghub.elsevier.com/retrieve/pii/S1673-8527(18)30117-6). 

## Result Folder structure

`Result` folder under current path(default) or output_folder set by user. A typical structure of `Result` is follows:

        Result/
            ├── QC
            │   ├── N1141_1.clean_fastqc.html
            │   ├── N1141_2.clean_fastqc.html
            │   ├── N1177_1.clean_fastqc.html
            │   └── N1177_2.clean_fastqc.html
            ├── Identified_lncRNA
            │   ├── all_lncRNA_for_classifier.gtf
            │   ├── final_all.fa
            │   ├── final_all.gtf
            │   ├── lncRNA.fa
            │   ├── protein_coding.fa
            │   └── protein_coding.final.gtf
            ├── LncReporter
            │   ├── Differential_Expression_analysis.csv
            │   └── Report.html
            ├── Quantification
            │   ├── kallisto.count.txt
            │   └── kallisto.tpm.txt
            └── Star_alignment
                ├── STAR_N1141
                │   ├── N1141Aligned.sortedByCoord.out.bam
                │   ├── N1141Log.final.out
                │   ├── N1141Log.out
                │   ├── N1141Log.progress.out
                │   └── N1141SJ.out.tab
                └── STAR_N1177
                    ├── N1177Aligned.sortedByCoord.out.bam
                    ├── N1177Log.final.out
                    ├── N1177Log.out
                    ├── N1177Log.progress.out
                    └── N1177SJ.out.tab

* `QC` stored the Quality control output generated by FastQC or AfterQC software.<br>
* `Identified_lncRNA` contains all assembled lncRNA and their sequences. *all_lncRNA_for_classifier.gtf* includes both novel and known lncRNA features in [GTF format](http://www.ensembl.org/info/website/upload/gff.html);
*lncRNA.fa* is all lncRNA sequences in fasta format. *protein_coding.final.gtf* and *protein_coding.fa* are protein coding information extracted from gencode annotation. *final_all.gtf* and *final_all.fa* are combined files for further analysis.<br>
* `Alignment` are hisat/tophat/STAR aligner standard output<br>
* `Quantification` are estimated abundance using kallisto. *kallisto.count.txt* stored reads count matrix and *kallisto.tpm.txt* are tpm(Transcripts Per Kilobase Million) matrix.
* `LncReporter` stored the interactive report file and differential expression matrix generated by LncPipeReporter which wrapped EdgeR.


**Output directory: `results/QC`**

* `sample_fastqc.html`
  * FastQC report, containing quality metrics for your untrimmed raw fastq files
* `zips/sample_fastqc.zip`
  * zip file containing the FastQC report, tab-delimited data file and plot images


**Output directory: `results/Identified_lncRNA`**  

**Output directory: `results/LncReporter`**  

**Output directory: `results/Identified_lncRNA`**  

**Output directory: `results/Quantification`**  

**Output directory: `results/*_alignment`**

